cirq-core/cirq/devices/noise_utils.py

```diff
@@ -78,20 +78,25 @@ def __str__(self):
         return f'{self.gate_type}{self.qubits}'
 
     def __repr__(self) -> str:
-        fullname = f'{self.gate_type.__module__}.{self.gate_type.__qualname__}'
         qubits = ', '.join(map(repr, self.qubits))
-        return f'cirq.devices.noise_utils.OpIdentifier({fullname}, {qubits})'
+        if hasattr(self.gate_type, '__qualname__'):
+            gate = f'{self.gate_type.__module__}.{self.gate_type.__qualname__}'
+        else:
+            gate = repr(self.gate_type)
+        return f'cirq.devices.noise_utils.OpIdentifier({gate}, {qubits})'
```

Comment:
Why the `if else`? Why can't we do 
```suggestion
        return f'cirq.devices.noise_utils.OpIdentifier({repr(self.gate_type)}, {qubits})'
```

Reply:
Because, for example, `repr(cirq.SingleQubitCliffordGate)` outputs `"<class 'cirq.ops.clifford_gate.SingleQubitCliffordGate'>"`, but the desired output is `"cirq.ops.clifford_gate.SingleQubitCliffordGate"`.

Reply:
```python
import cirq

cirq.testing.assert_equivalent_repr(cirq.SingleQubitCliffordGate)
```

fails. Let's open a different PR that fixes this.

Reply:
Actually maybe not :thinking: . This is not an instance. It's a class. Maybe the `repr` is right after all..

Reply:
Yes, I think this is the intended behavior of `repr`, and the `if else` is needed to handle the case where a class is passed to `OpIdentifier` vs the case where an instance is passed.

Reply:
Oh 

```python
import cirq

eval("cirq.SingleQubitCliffordGate")
```

passes so this is broken after all.

Reply:
So `repr(cirq.SingleQubitCliffordGate)` should output `"cirq.SingleQubitCliffordGate"`? I'm not sure how to define `repr` for this as it is not an instance of the object.

Reply:
You are right. We can't define `repr` for this since it's not an instance. This is something we expect [proper_repr](https://github.com/quantumlib/Cirq/blob/8e4e7d147d6d4ea5dd8111ba21a715aa3acd955c/cirq-core/cirq/_compat.py#L135) to handle.

We can add something like

```python
    if hasattr(value, "__qualname__"):
        return f"{value.__module__}.{value.__qualname__}"
```

in `proper_repr` and call `proper_repr` here.

Reply:
Added.